internal/types/compare.go

```diff
@@ -255,3 +254,130 @@ func compareNumbers(a float64, b int64) CompareResult {
 
 	return CompareResult(bigA.Cmp(bigB))
 }
+
+// dataTypeOrderResult represents the comparison order of data types.
+type dataTypeOrderResult uint8
+
+const (
+	_ dataTypeOrderResult = iota
+	nullDataType
+	nanDataType
+	numbersDataType
+	stringDataType
+	objectDataType
+	arrayDataType
+	binDataType
+	objectIdDataType
+	booleanDataType
+	dateDataType
+	timestampDataType
+	regexDataType
+)
+
+// defineDataType define which type has value and returns a sequence the type has.
+func defineDataType(value any) dataTypeOrderResult {
```

Comment:
As @AlekSi mentioned, `detectDataType` would be a better name here.

Reply:
done